pkg/sentry/kernel/msgqueue/msgqueue.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package msgqueue implements System V message queues.
+package msgqueue
+
+import (
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/sentry/fs"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/ipc"
+	ktime "gvisor.dev/gvisor/pkg/sentry/kernel/time"
+	"gvisor.dev/gvisor/pkg/sync"
+	"gvisor.dev/gvisor/pkg/syserror"
+	"gvisor.dev/gvisor/pkg/waiter"
+)
+
+const (
+	// System-wide limit for maximum number of queues.
+	maxQueues = linux.MSGMNI
+
+	// Maximum size of a queue in bytes.
+	maxQueueBytes = linux.MSGMNB
+
+	// Maximum size of a message in bytes.
+	maxMessageBytes = linux.MSGMAX
+)
+
+// Registry contains a set of message queues that can be referenced using keys
+// or IDs.
+//
+// +stateify savable
+type Registry struct {
+	// userNS owning the IPC namespace this registry belongs to. Immutable.
+	userNS *auth.UserNamespace
+
+	// mu protects all the fields below.
+	mu sync.Mutex `state:"nosave"`
```

Comment:
Don't need this anymore, all fields are immutable after creation. In fact the whole Registry type can probably become a newtype definition on `ipc.Registry`.

Reply:
We need a `Mutex` to protect `reg`. I updated `ipc.Registry`'s comment to be a bit clearer about this. The reason this is done instead of including the mutex inside `ipc.Registry` is because `shm` and `semaphore` need to protect other fields in their registry implementation, so I thought it'd be better to use only one mutex per registry, instead of having one inside `ipc.Registry` and another in `[shm/semaphore].Registry`.

Reply:
Per external discussion, `ipc.Registry` is not threadsafe and the caller is responsible for synchronization. We don't want to move the mutex into `ipc.Registry` because the caller needs a broader critical section around registry calls anyways.